substrate-node/pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -534,68 +675,163 @@ use sp_std::convert::{TryFrom, TryInto};
 use tfchain_support::types::PublicIP;
 // Internal functions of the pallet
 impl<T: Config> Pallet<T> {
-    pub fn _create_node_contract(
-        account_id: T::AccountId,
+    pub fn _create_group(account_id: T::AccountId) -> DispatchResultWithPostInfo {
+        let twin_id = pallet_tfgrid::TwinIdByAccountID::<T>::get(&account_id)
+            .ok_or(Error::<T>::TwinNotExists)?;
+
+        let mut id = GroupID::<T>::get();
+        id = id + 1;
+
+        let new_group = types::Group {
+            id: id,
+            twin_id: twin_id,
+            capacity_reservation_contract_ids: vec![],
+        };
+
+        Groups::<T>::insert(id, &new_group);
+
+        Self::deposit_event(Event::GroupCreated {
+            group_id: id,
+            twin_id: twin_id,
+        });
+
+        Ok(().into())
+    }
+
+    pub fn _delete_group(account_id: T::AccountId, group_id: u32) -> DispatchResultWithPostInfo {
+        let twin_id = pallet_tfgrid::TwinIdByAccountID::<T>::get(&account_id)
+            .ok_or(Error::<T>::TwinNotExists)?;
+
+        let group = Groups::<T>::get(group_id).ok_or(Error::<T>::GroupNotExists)?;
+
+        ensure!(
+            twin_id == group.twin_id,
+            Error::<T>::TwinNotAuthorizedToDeleteGroup
+        );
+        ensure!(
+            group.capacity_reservation_contract_ids.is_empty(),
+            Error::<T>::GroupHasActiveMembers
+        );
+
+        Groups::<T>::remove(group_id);
+
+        Self::deposit_event(Event::GroupDeleted { group_id: group_id });
+
+        Ok(().into())
+    }
+
+    pub fn _add_capacity_reservation_contract_to_group(
+        group_id: u32,
+        capacity_rservation_id: u64,
         node_id: u32,
-        deployment_hash: DeploymentHash,
-        deployment_data: DeploymentDataInput<T>,
-        public_ips: u32,
+    ) -> DispatchResultWithPostInfo {
+        let mut group = Groups::<T>::get(group_id).ok_or(Error::<T>::GroupNotExists)?;
+        group
+            .capacity_reservation_contract_ids
+            .push(capacity_rservation_id);
+        Groups::<T>::insert(group_id, &group);
+
+        CapacityReservationIDByNodeGroupConfig::<T>::insert(
+            types::NodeGroupConfig {
+                group_id: group_id,
```

Comment:
shorthand notation here?

Reply:
Why? I think it's always better to use this notation to avoid mistakes later (due to git merging). For example lets say we introduce a new variable for that object and insert it in between two others. The order could be wrong after a mergeconflict that is automatically resolved. If those two types are the same the code will build and it will introduce a bug.